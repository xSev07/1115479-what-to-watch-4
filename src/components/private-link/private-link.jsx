import React from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {AppRoute} from "../../const";
import {connect} from "react-redux";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import {AuthorizationStatus} from "../../reducer/user/user";
import {extendObject} from "../../utils/common/common";

const PrivateLink = (props) => {
  const {userAuthorized, className, children} = props;
  const attributes = extendObject({}, props);
  delete attributes.userAuthorized;
  delete attributes.children;
  delete attributes.dispatch;

  return (
    userAuthorized ?
      <button
        {...attributes}
      >
        {children}
      </button>
      :
      <Link
        to={AppRoute.LOGIN}
        className={className}
      >
        {children}
      </Link>
  );
};

PrivateLink.propTypes = {
  userAuthorized: PropTypes.bool.isRequired,
  className: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
  ]),
};

const mapStateToProps = (state) => ({
  userAuthorized: getAuthorizationStatus(state) === AuthorizationStatus.AUTH,
});

export {PrivateLink};
export default connect(mapStateToProps)(PrivateLink);
