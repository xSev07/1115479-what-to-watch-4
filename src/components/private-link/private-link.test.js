import React from "react";
import renderer from "react-test-renderer";
import {PrivateLink} from "./private-link";
import {StaticRouter} from "react-router-dom";

describe(`Should PrivateLink render correctly`, () => {
  it(`should render button`, () => {
    const tree = renderer
      .create(
          <PrivateLink
            className={`test`}
            userAuthorized={true}
          >
            <span>Button</span>
          </PrivateLink>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render link`, () => {
    const tree = renderer
      .create(
          <StaticRouter>
            <PrivateLink
              className={`test`}
              userAuthorized={false}
            >
              <span>Link</span>
            </PrivateLink>
          </StaticRouter>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
