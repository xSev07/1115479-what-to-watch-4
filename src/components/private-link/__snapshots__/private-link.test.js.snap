// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should PrivateLink render correctly should render button 1`] = `
<button
  className="test"
>
  <span>
    Button
  </span>
</button>
`;

exports[`Should PrivateLink render correctly should render link 1`] = `
<a
  className="test"
  href="/login"
  onClick={[Function]}
>
  <span>
    Link
  </span>
</a>
`;
